modules/auxiliary/gather/ldap_esc_vulnerable_cert_finder.rb

```diff
@@ -315,6 +353,142 @@ def find_esc3_vuln_cert_templates
     query_ldap_server_certificates(esc3_template_2_raw_filter, 'ESC3_TEMPLATE_2')
   end
 
+  def find_esc4_vuln_cert_templates
+    # Determine who we are authenticating with. Retrieve the username and user SID
+    whoami_response = ''
+    begin
+      whoami_response = @ldap.ldapwhoami
+    rescue Net::LDAP::Error => e
+      print_warning("The module failed to run the ldapwhoami command, ESC4 detection can't continue. Error was: #{e.class}: #{e.message}.")
+      return
+    end
+
+    if whoami_response.empty?
+      print_error("Unable to retrieve the username using ldapwhoami, ESC4 detection can't continue")
+      return
+    end
+
+    sam_account_name = whoami_response.split('\\')[1]
+    user_raw_filter = "(sAMAccountName=#{sam_account_name})"
+    attributes = ['objectSID', 'DN']
+    our_user = query_ldap_server(user_raw_filter, attributes)&.first
+    if our_user.nil?
+      print_warning("Unable to determine the User SID for #{sam_account_name}, ESC4 detection can't continue")
+      return
+    else
+      user_sid = Rex::Proto::MsDtyp::MsDtypSid.read(our_user[:objectsid].first).value
+    end
+
+    # Determine if we are a User or a Computer
+    fiter_user_or_computer = "(objectSid=#{user_sid})"
+    attributes = ['objectClass']
+    response_computer_or_user = query_ldap_server(fiter_user_or_computer, attributes)&.first
+
+    if response_computer_or_user.nil?
+      print_warning("Unable to determine the if SID belongs to a user account or a computer account, ESC4 detection can't continue")
+      return
+    end
+
+    if response_computer_or_user[:objectclass].include?('computer')
+      computer_or_user = 'computer'
+    else
+      computer_or_user = 'user'
+    end
+
+    # Get the domain SID (from the user_sid which is domain_sid + rid)
+    domain_sid = ''
+    if user_sid =~ /(S-[0-9]*-[0-9]*-[0-9]*-[0-9]*-[0-9]*-[0-9]*)/
+      domain_sid = Regexp.last_match(1)
+    else
+      print_warning("Unable to derive the Domain SID from the User SID, ESC4 detection can't continue")
+      return
+    end
+
+    user_groups = []
+    if computer_or_user == 'computer'
+      user_groups << "#{domain_sid}-#{Rex::Proto::Secauthz::WellKnownSids::DOMAIN_GROUP_RID_COMPUTERS}"
+    else
+      user_groups << "#{domain_sid}-#{Rex::Proto::Secauthz::WellKnownSids::DOMAIN_GROUP_RID_USERS}"
+      user_groups << Rex::Proto::Secauthz::WellKnownSids::SECURITY_AUTHENTICATED_USER_SID
+    end
+
+    # Perform an LDAP query to get the groups the user is a part of
+    # Use LDAP_MATCHING_RULE_IN_CHAIN OID in order to walk the chain of ancestry of groups.
+    # https://learn.microsoft.com/en-us/windows/win32/adsi/search-filter-syntax?redirectedfrom=MSDN
+    filter_with_user = <<~FILTER
+      (&(member:1.2.840.113556.1.4.1941:=CN=#{our_user[:dn].first}))
+    FILTER
```

Comment:
So to handle cases where `Domain Users` for example is a member of a group with write access, we'll want to split this out and search where the member is our user, or one of the groups that they're implicitly in as defined in `user_groups`.

```suggestion
    filter_with_user = "(|(member:1.2.840.113556.1.4.1941:=#{our_user[:dn].first})"
    user_groups.each do |sid|
      obj = query_ldap_server("(objectSid=#{sid})", ['dn'])&.first
      print_error("Failed to lookup SID.") unless obj
      filter_with_user << "(member:1.2.840.113556.1.4.1941:=#{obj[:dn].first})" if obj
    end
    filter_with_user << ')'
```

This also changes from the logical and operator `(&` to the logical or `(|` so we can check for any of the SIDs.